pkg/prometheus/promcfg.go

```diff
@@ -586,6 +613,177 @@ func (cg *configGenerator) generatePodMonitorConfig(
+		// Exact label matches.
+		for k := range m.Spec.Targets.Ingress.Selector.MatchLabels {
+			labelKeys = append(labelKeys, k)
+		}
+		sort.Strings(labelKeys)
+
+		for _, k := range labelKeys {
+			relabelings = append(relabelings, yaml.MapSlice{
+				{Key: "action", Value: "keep"},
+				{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(k)}},
+				{Key: "regex", Value: m.Spec.Targets.Ingress.Selector.MatchLabels[k]},
+			})
+		}
+
+		// Set based label matching. We have to map the valid relations
+		// `In`, `NotIn`, `Exists`, and `DoesNotExist`, into relabeling rules.
+		for _, exp := range m.Spec.Targets.Ingress.Selector.MatchExpressions {
+			switch exp.Operator {
+			case metav1.LabelSelectorOpIn:
+				relabelings = append(relabelings, yaml.MapSlice{
+					{Key: "action", Value: "keep"},
+					{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(exp.Key)}},
+					{Key: "regex", Value: strings.Join(exp.Values, "|")},
+				})
+			case metav1.LabelSelectorOpNotIn:
+				relabelings = append(relabelings, yaml.MapSlice{
+					{Key: "action", Value: "drop"},
+					{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(exp.Key)}},
+					{Key: "regex", Value: strings.Join(exp.Values, "|")},
+				})
+			case metav1.LabelSelectorOpExists:
+				relabelings = append(relabelings, yaml.MapSlice{
+					{Key: "action", Value: "keep"},
+					{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(exp.Key)}},
+					{Key: "regex", Value: ".+"},
+				})
+			case metav1.LabelSelectorOpDoesNotExist:
+				relabelings = append(relabelings, yaml.MapSlice{
+					{Key: "action", Value: "drop"},
+					{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(exp.Key)}},
+					{Key: "regex", Value: ".+"},
+				})
+			}
+		}
+
+		if version.Major == 1 && version.Minor < 7 {
+			if apiserverConfig != nil {
+				level.Info(cg.logger).Log("msg", "custom apiserver config is set but it will not take effect because prometheus version is < 1.7")
+			}
+			cfg = append(cfg, cg.generateK8SSDConfig(nil, nil, nil, kubernetesSDRoleIngress))
+		} else {
+			selectedNamespaces := getNamespacesFromNamespaceSelector(&m.Spec.Targets.Ingress.NamespaceSelector, m.Namespace, ignoreNamespaceSelectors)
+			cfg = append(cfg, cg.generateK8SSDConfig(selectedNamespaces, apiserverConfig, basicAuthSecrets, kubernetesSDRoleIngress))
+		}
+
+		// Relabelings for blackbox exporter.
+		relabelings = append(relabelings, []yaml.MapSlice{
+			{
+				{Key: "source_labels", Value: "[__address__]"},
+				{Key: "target_label", Value: "__param_target"},
+			},
+			{
+				{Key: "source_labels", Value: "[__param_target]"},
+				{Key: "target_label", Value: "instance"},
+			},
+			{
+				{Key: "target_label", Value: "__address__"},
+				{Key: "replacement", Value: m.Spec.BlackboxExporter.URL},
+			},
```

Comment:
Shouldn't we add a label “probe” containing the blackbox exporter address so that we can easily differentiate metrics coming from differents blackbox exporter instances?

Reply:
Sure. Would be possible. But I'd rather avoid adding more stuff and discussion to this PR - though this is minor. How do you feel about that @brancz?

Reply:
I prefer to start minimal, adding something later is easier than removing :)